c2rust-analyze/src/panic_detail.rs

```diff
@@ -0,0 +1,170 @@
+        msg: panic_to_string(info.payload()),
+        loc: info.location().map(|l| l.to_string()),
+        relevant_loc: guess_relevant_loc(&bt),
+        backtrace: Some(bt),
+        sp: CURRENT_SPAN.with(|cell| cell.get()),
+    };
+    let old = CURRENT_PANIC_DETAIL.with(|cell| cell.replace(Some(detail)));
+    if let Some(old) = old {
+        warn!("discarding old panic detail: {:?}", old);
+    }
+}
+
+/// Get the [`PanicDetail`] of the most recent panic.  This clears the internal storage, so if this
+/// is called twice in a row without an intervening panic, the second call always returns `None`.
+pub fn take_current() -> Option<PanicDetail> {
+    CURRENT_PANIC_DETAIL.with(|cell| cell.take())
+}
+
+/// Get the [`PanicDetail`] of the most recent panic; if it's not available, build a placeholder
+/// `PanicDetail` from the panic payload `e`.  This is useful in conjunction with
+/// [`std::panic::catch_unwind`].
+pub fn catch(e: &(dyn Any + Send + 'static)) -> PanicDetail {
+    take_current().unwrap_or_else(|| {
+        let msg = panic_to_string(e);
+        warn!("missing panic detail; caught message {:?}", msg);
+        PanicDetail::new(msg)
+    })
+}
+
+/// Crude heuristic to guess the first interesting location in a [`Backtrace`], skipping over
+/// helper functions, wrappers, and panic machinery.  The resulting location is used in the summary
+/// message produced by [`PanicDetail::to_string_short`].
+fn guess_relevant_loc(bt: &Backtrace) -> Option<String> {
+    for frame in bt.frames() {
+        for symbol in frame.symbols() {
+            let name = match symbol.name() {
+                Some(x) => x.to_string(),
+                None => continue,
+            };
+            if name.starts_with("c2rust_analyze::dataflow")
+                || name.starts_with("c2rust_analyze::borrowck")
+                || name.starts_with("c2rust_analyze::rewrite")
+                || name.contains("type_of_rvalue")
+                || name.contains("TypeOf")
+                || name.contains("lty_project")
+            {
+                let filename_str = match symbol.filename() {
+                    Some(x) => x.display().to_string(),
+                    None => "[unknown]".to_string(),
+                };
+                return Some(format!(
+                    "{} @ {}:{}:{}",
+                    name,
+                    filename_str,
+                    symbol.lineno().unwrap_or(0),
+                    symbol.colno().unwrap_or(0)
+                ));
+            }
+        }
+    }
+    None
+}
+
+fn panic_to_string(e: &(dyn Any + Send + 'static)) -> String {
+    match e.downcast_ref::<&'static str>() {
+        Some(s) => return s.to_string(),
+        None => {}
+    }
+
+    match e.downcast_ref::<String>() {
+        Some(s) => return (*s).clone(),
+        None => {}
+    }
+
+    format!("unknown error: {:?}", e.type_id())
+}
+
+thread_local! {
+    static CURRENT_SPAN: Cell<Span> = Cell::new(DUMMY_SP);
```

Comment:
```suggestion
    static CURRENT_SPAN: Cell<Span> = Default::default;
```

Reply:
I don't see much value in doing it this way.